src/main/java/build/buildfarm/instance/AbstractServerInstance.java

```diff
@@ -434,7 +438,11 @@ protected void updateOperationWatchers(Operation operation) {
 
   @Override
   public Operation getOperation(String name) {
-    return outstandingOperations.get(name);
+    Operation operation = completedOperations.get(name);
+    if (operation == null) {
```

Comment:
There's a possible race here. If there is a concurrent call to updateOperationWatchers, which removes an operation from outstandingOperations and adds it to completedOperations, then a call to this method may be unable to find the operation. I understand that's why you're doing a put before the remove above, but I'm not sure that the Jvm guarantees that the writes complete in that order. In any case, this subtlety should be documented in the code.